lib/cfa/augeas_parser/writer.rb

```diff
@@ -0,0 +1,155 @@
+    # no entry in tree, then do not insert position, which means, that
+    # following setting of value appends it to the end.
+    #
+    # @param last [String] path of last valid entry written to tree
+    # @param key [String] key that need to be inserted
+    # @param tree [CFA::AugeasTree] tree where is key located
+    def insert_entry(last, key, tree)
+      if last
+        report_error { aug.insert(last, key, false) }
+      else
+        e = find_first_valid(tree.data(filtered: false))
+        if e
+          e_path = prefix + "/" + e[:orig_key]
+          report_error { aug.insert(e_path, key, true) }
+        end
+      end
+    end
+
+    # Do modification according to operation defined in AugeasEntry
+    # @param tree [CFA::AugeasTree] tree where entry lives
+    # @param entry [AugeasElement] entry to process
+    # @param last_valid_entry_path [String, nil] path of last valid entry
+    #   written or nil if there is no such entry
+    # @param key [String] valid augeas key for entry. Important for collections
+    #   which have key without its index, but augeas do not allow it.
+    # @param path [String] whole path where to write entry in augeas
+    def process_operation(tree, entry,
+      last_valid_entry_path, key, path)
+      case entry[:operation]
+      when :add, nil # add is default operation
+        insert_entry(last_valid_entry_path, key, tree)
+        set_entry(path, entry)
+      when :remove then report_error { aug.rm(path) }
+      when :modify then set_entry(path, entry[:value])
+      when :keep then recurse_write(path, entry)
+      else raise "invalid operation"
+      end
+    end
+
+    # writes value of entry to path and if it have sub-tree then call {write} on it
+    # @param path [String] path where to write
+    # @param entry [AugeasElement] entry to write
+    def set_entry(path, entry)
+      value = entry[:value]
+      case value
+      when AugeasTree
+        report_error { aug.touch(path) }
+      when AugeasTreeValue
+        report_error { aug.set(path, value.value) }
+      else
+        report_error { aug.set(path, value) }
+      end
+      recurse_write(path, entry)
+    end
+
+    # calls write on entry if entry have sub-tree
+    # @param path [String] path of entry
+    # @param entry [AugeasElement]
+    def recurse_write(path, entry)
+      if entry[:value].is_a?(AugeasTree)
+        write(path, entry[:value])
+      elsif entry[:value].is_a?(AugeasTreeValue)
+        write(path, entry[:value].tree)
+      end
+    end
+
+    # Finds first entry that is already in augeas, so it means entry,
+    # that is kept or only modified
+    def find_first_valid(data)
+      data.find { |e| [:keep, :modify].include?(e[:operation]) }
+    end
+
+    # Calls block and if it failed, raise exception with details from augeas
+    # why it failed
+    # @yield call to aug that is secured
+    # @raise [RuntimeError]
+    def report_error
+      return if yield
+
+      error = aug.error
```

Comment:
ok, this cannot happen, I remove all report_error calls. As majority of aug call return int and not booleand and aug.set cannot fail as it just set value to tree